src/main/java/workflow/context/AbstractCompositeContext.java

```diff
@@ -19,10 +19,12 @@
      * Instantiate a Context with a State
      * @param state ProcessState
      */
-    AbstractCompositeContext(ProcessState state, ContextInterface parentContext) {
+    AbstractCompositeContext(ProcessState state, ContextInterface parentContext, String key) {
         this.state = state;
         this.parent = parentContext;
-        this.childContexts = new HashMap<String, ContextInterface>();//TODO: should be a dictionary? will require streams work
+        this.key = key;
+        this.childContexts = new HashMap<String, ContextInterface>();
```

Comment:
The keys should all actually be aware of their relationships to each other, every key knows its parent and its children.